base/strings/annotated_render.jl

```diff
@@ -0,0 +1,185 @@
+    end
+end
+
+write(io::IO, s::Union{<:AnnotatedString, SubString{<:AnnotatedString}}) =
+    _ansi_writer(io, s, write)::Int
+
+print(io::IO, s::Union{<:AnnotatedString, SubString{<:AnnotatedString}}) =
+    (_ansi_writer(io, s, print); nothing)
+
+# We need to make sure that printing to an `AnnotatedIOBuffer` calls `write` not `print`
+# so we get the specialised handling that `_ansi_writer` doesn't provide.
+print(io::AnnotatedIOBuffer, s::Union{<:AnnotatedString, SubString{<:AnnotatedString}}) =
+    (write(io, s); nothing)
+
+escape_string(io::IO, s::Union{<:AnnotatedString, SubString{<:AnnotatedString}},
+              esc = ""; keep = ()) =
+    (_ansi_writer(io, s, (io, s) -> escape_string(io, s, esc; keep)); nothing)
+
+function write(io::IO, c::AnnotatedChar)
+    if get(io, :color, false) == true
+        active_style::Any = nothing # Represents the currently-applied style
+
+        # TODO: re-factor into separate (shared) function
+        link = nothing
+        merged_style = nothing
+        for (key, style) in reverse(c.annotations)
+            if key === :link
+                link = style::String
+            end
+            key !== :face && continue
+            # Merge as many of these as we can into a single style before
+            # we print to the terminal
+            (merged_style, successful) = mergestyle_(merged_style, style)
+            if !successful
+                active_style = termstyle_(buf, merged_style, active_style)
+                merged_style = style
+            end
+        end
+
+        active_style = termstyle_(io, merged_style, active_style)
+        bytes = write(io, c.char)
+        termreset_(io, active_style)
+        bytes
+    else
+        write(io, c.char)
+    end
+end
+
+print(io::IO, c::AnnotatedChar) = (write(io, c); nothing)
+
+function show(io::IO, c::AnnotatedChar)
+    if get(io, :color, false) == true
+        out = IOBuffer()
+        show(out, c.char)
+        cstr = AnnotatedString(
+            String(take!(out)[2:end-1]),
+            [(1:ncodeunits(c), a...) for a in c.annotations])
+        print(io, ''', cstr, ''')
+    else
+        show(io, c.char)
+    end
+end
+
+function write(io::IO, aio::AnnotatedIOBuffer)
+    if get(io, :color, false) == true
+        # This does introduce an overhead that technically
+        # could be avoided, but I'm not sure that it's currently
+        # worth the effort to implement an efficient version of
+        # writing from a AnnotatedIOBuffer with style.
+        # In the meantime, by converting to an `AnnotatedString` we can just
+        # reuse all the work done to make that work.
+        write(io, read(aio, AnnotatedString))
+    else
+        write(io, aio.io)
+    end
+end
+
+function show(io::IO, ::MIME"text/html", s::Union{<:AnnotatedString, SubString{<:AnnotatedString}})
+    htmlescape(str) = replace(str, '&' => "&amp;", '<' => "&lt;", '>' => "&gt;")
```

Comment:
This seems incomplete, since the commonly given list is `&, <, ', ", \0` https://cheatsheetseries.owasp.org/cheatsheets/Cross_Site_Scripting_Prevention_Cheat_Sheet.html#output-encoding-rules-summary